cli/flox/src/commands/publish.rs

```diff
@@ -0,0 +1,134 @@
+use anyhow::{bail, Result};
+use bpaf::Bpaf;
+use flox_rust_sdk::flox::Flox;
+use flox_rust_sdk::models::lockfile::Lockfile;
+use flox_rust_sdk::providers::build::FloxBuildMk;
+use flox_rust_sdk::providers::publish::{
+    check_build_metadata,
+    check_environment_metadata,
+    Publish,
+    PublishProvider,
+};
+use indoc::indoc;
+use log::debug;
+use tracing::instrument;
+
+use super::{environment_select, EnvironmentSelect};
+use crate::commands::{ensure_floxhub_token, ConcreteEnvironment};
+use crate::config::Config;
+use crate::subcommand_metric;
+use crate::utils::message;
+
+#[allow(unused)] // remove when we implement the command
+#[derive(Bpaf, Clone)]
+pub struct PublishCmd {
+    #[bpaf(external(environment_select), fallback(Default::default()))]
+    environment: EnvironmentSelect,
+
+    #[bpaf(external(subcommand_or_build_targets))]
+    subcommand_or_targets: SubcommandOrBuildTargets,
+}
+
+#[derive(Debug, Bpaf, Clone)]
+enum SubcommandOrBuildTargets {
+    BuildTarget {
+        /// The package to build.
+        /// Corresponds to entries in the 'build' table in the environment's manifest.toml.
+        /// If not specified, all packages are built.
+        #[bpaf(positional("package"))]
+        target: String,
+    },
+}
+
+impl PublishCmd {
+    pub async fn handle(self, config: Config, flox: Flox) -> Result<()> {
+        if !config.features.unwrap_or_default().publish {
+            message::plain("🚧 👷 heja, a new command is in construction here, stay tuned!");
+            bail!("'publish' feature is not enabled.");
+        }
+
+        match self.subcommand_or_targets {
+            SubcommandOrBuildTargets::BuildTarget { target } => {
+                let env = self
+                    .environment
+                    .detect_concrete_environment(&flox, "Clean build files of")?;
+
+                Self::publish(flox, env, target).await
+            },
+        }
+    }
+
+    #[instrument(name = "publish", skip_all, fields(package))]
+    async fn publish(mut flox: Flox, env: ConcreteEnvironment, package: String) -> Result<()> {
+        subcommand_metric!("publish");
+
+        if let ConcreteEnvironment::Remote(_) = &env {
+            bail!("Cannot publish from a remote environment");
+        };
+
+        let mut env = env.into_dyn_environment();
+
+        if !check_package(&env.lockfile(&flox)?, &package)? {
+            bail!("Package '{}' not found in environment", package);
+        }
+
+        let (env_meta, build_meta) = (
+            check_environment_metadata(&flox, &*env).unwrap(),
+            check_build_metadata(&*env, &package, &flox.system).unwrap(),
+        );
```

Comment:
nonblocking question: would it be better for testing if this was inside `publish_provider.publish`? Not sure, just wondering

Reply:
I've got some open questions around the provider structure that may be worth a discussion with you guys.  It kinda evolved into this, butt worth another look as a follow up.